File: solutions/problem_1561_maximum_number_of_coins_you_can_get/solution.py
"""
Problem 1561: Maximum Number of Coins You Can Get
===============================================
Difficulty: Medium
Tags: Array, Math, Greedy, Sorting, Game Theory
Pattern: Greedy with Sorting

Time Complexity:  O(n log n)
Space Complexity: O(n)
"""

class Solution:
    def maxCoins(self, piles: List[int]) -> int:
        # Sort + greedy - O(n log n) time
        piles.sort()
        result = 0
        curr_end = 0
        for item in piles:
            if isinstance(item, (list, tuple)):
                if item[0] >= curr_end:
                    result += 1
                    curr_end = item[1]
            else:
                result += 1
        return result

File: solutions/problem_1561_maximum_number_of_coins_you_can_get/solution.cpp
/*
 * Problem 1561: Maximum Number of Coins You Can Get
 * ===============================================
 * Difficulty: Medium
 * Tags: Array, Math, Greedy, Sorting, Game Theory
 * Pattern: Greedy with Sorting
 *
 * Time Complexity:  O(n log n)
 * Space Complexity: O(n)
 */

#include <algorithm>
#include <string>
#include <vector>
using namespace std;

class Solution {
public:
    int maxCoins(vector<int>& piles) {
        // Sort + greedy - O(n log n) time
        sort(piles.begin(), piles.end());
        int result = 0, curr_end = 0;
        for (auto& item : piles) {
            result++;
        }
        return result;
    }
};
